cpp/src/strings/regex/regex.cuh

```diff
@@ -0,0 +1,184 @@
+* limitations under the License.
+*/
+#pragma once
+
+#include <cuda_runtime.h>
+#include <memory>
+#include <functional>
+
+namespace cudf
+{
+
+class string_view;
+
+namespace strings
+{
+namespace detail
+{
+
+struct Reljunk;
+struct Reinst;
+class Reprog;
+
+/**
+ * @brief Regex class stored on the device and executed by Reprog_device.
+ *
+ * This class holds the unique data for any regex CCLASS instruction.
+ */
+class Reclass_device
+{
+public:
+    int32_t builtins{};
+    int32_t count{};
+    char32_t* literals{};
+
+    __device__ bool is_match(char32_t ch, const uint8_t* flags);
+};
+
+/**
+ * @brief Regex program of instructions/data for a specific regex pattern.
+ *
+ * Once create, this find/extract methods are used to evaluating the regex instructions
+ * against a single string.
+ */
+class Reprog_device
+{
+public:
+    Reprog_device() = delete;
+    ~Reprog_device() = default;
+    Reprog_device(const Reprog_device&) = default;
+    Reprog_device(Reprog_device&&) = default;
+    Reprog_device& operator=(const Reprog_device&) = default;
+    Reprog_device& operator=(Reprog_device&&) = default;
+
+    /**
+     * @brief Create device program instance from a regex pattern.
+     *
+     * The number of strings is needed to compute the stack size required when executing the regex.
+     *
+     * @param pattern The regex pattern to compile.
+     * @param cp_flags The code-point lookup table for character types.
+     * @param strings_count Number of strings that will be evaluated.
+     * @param stream CUDA stream for asynchronous memory allocations.
+     * @return The program device object.
+     */
+    static std::unique_ptr<Reprog_device, std::function<void(Reprog_device*)>>
+        create(std::string const& pattern, const uint8_t* cp_flags, int32_t strings_count, cudaStream_t stream=0);
+    /**
+     * @brief Called automatically by the unique_ptr returned from create().
+     */
+    void destroy();
+
+    /**
+     * @brief Returns the number of regex instructions.
+     */
+    int32_t insts_counts()   { return _insts_count; }
+
+    /**
+     * @brief Returns the number of regex groups found in the expression.
+     */
+    int32_t group_counts()   { return _num_capturing_groups; }
+
+    /**
+     * @brief This sets up the memory used for keeping track of the regex progress.
+     *
+     * Call this for each string before calling find or extract.
+     */
+    __device__ inline void set_stack_mem(u_char* s1, u_char* s2);
+
+    /**
+     * @brief Returns the regex instruction object for a given index.
+     */
+    __host__ __device__ inline Reinst* get_inst(int32_t idx);
+
+    /**
+     * @brief Returns the regex class object for a given index.
+     */
+    __device__ inline Reclass_device get_class(int32_t idx);
+
+    /**
+     * @brief Returns the start-instruction-ids vector.
+     */
+    __device__ inline int32_t* get_startinst_ids();
+
+    /**
```

Comment:
In general, I prefer dropping the `get_` from accessor methods. Is this something we can do?

Reply:
I'd like to keep `get_inst` and `get_class` which take index arguments.

Reply:
Aside: please don't unnecessarily abbreviate. The meaning of `inst` is ambiguous.